homeassistant/components/bluetooth/passive_update_coordinator.py

```diff
@@ -0,0 +1,304 @@
+    async_address_present,
+    async_register_callback,
+)
+from .const import DOMAIN
+
+UNAVAILABLE_SECONDS = 60 * 5
+NEVER_TIME = -UNAVAILABLE_SECONDS
+
+
+@dataclasses.dataclass(frozen=True)
+class PassiveBluetoothEntityKey:
+    """Key for a passive bluetooth entity.
+
+    Example:
+    key: temperature
+    device_id: outdoor_sensor_1
+    """
+
+    key: str
+    device_id: str | None
+
+
+_T = TypeVar("_T")
+
+
+@dataclasses.dataclass(frozen=True)
+class PassiveBluetoothDataUpdate(Generic[_T]):
+    """Generic bluetooth data."""
+
+    devices: dict[str | None, DeviceInfo] = dataclasses.field(default_factory=dict)
+    entity_descriptions: dict[
+        PassiveBluetoothEntityKey, EntityDescription
+    ] = dataclasses.field(default_factory=dict)
+    entity_data: dict[PassiveBluetoothEntityKey, _T] = dataclasses.field(
+        default_factory=dict
+    )
+
+
+_PassiveBluetoothDataUpdateCoordinatorT = TypeVar(
+    "_PassiveBluetoothDataUpdateCoordinatorT",
+    bound="PassiveBluetoothDataUpdateCoordinator[Any]",
+)
+
+
+class PassiveBluetoothDataUpdateCoordinator(Generic[_T]):
+    """Bluetooth data update coordinator."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        logger: logging.Logger,
+        address: str,
+        update_method: Callable[[BluetoothServiceInfo], PassiveBluetoothDataUpdate[_T]],
+    ) -> None:
+        """Initialize the coordinator."""
+        self.hass = hass
+        self.logger = logger
+        self.name: str | None = None
+        self.address = address
+        self._listeners: list[
+            Callable[[PassiveBluetoothDataUpdate[_T] | None], None]
+        ] = []
+        self._entity_key_listeners: dict[
+            PassiveBluetoothEntityKey | None,
+            list[Callable[[PassiveBluetoothDataUpdate[_T] | None], None]],
+        ] = {}
+        self.update_method = update_method
+
+        self.entity_data: dict[PassiveBluetoothEntityKey, _T] = {}
+        self.entity_descriptions: dict[
+            PassiveBluetoothEntityKey, EntityDescription
+        ] = {}
+        self.devices: dict[str | None, DeviceInfo] = {}
+
+        self.last_update_success = True
+        self._last_callback_time: float = NEVER_TIME
+        self._present = True
+
+    @property
+    def available(self) -> bool:
+        """Return if the device is available."""
+        return self._present and self.last_update_success
+
+    def _async_check_device_present(self, _: datetime) -> None:
+        """Check if the device is present."""
+        if (
+            not self._present
+            or time.monotonic() - self._last_callback_time < UNAVAILABLE_SECONDS
+            or async_address_present(self.hass, self.address)
+        ):
+            return
+        self._present = False
+        self.async_update_listeners(None)
+
+    @callback
+    def async_setup(self) -> CALLBACK_TYPE:
+        """Start the callback."""
+        cancels = [
+            async_track_time_interval(
```

Comment:
Instead of listening to an interval, shouldn't this be using a track point in time that is set to "last update" + `UNAVAILABLE_SECONDS` ?

Reply:
With the coordinator in Shelly we do something like this by having the update interval be used as a fallback while relying on push notifications being pushed using `async_set_data` (that's on a normal coordinator)

Reply:
Probably need a guard for hass stopping as well

Reply:
17991a9564055fe6ff366cdf938c0f6df5cdea74

Reply:
Looks like we need to change it back to the track time interval because we are checking `async_address_present` as well which is also more expensive than I had hoped so I'll need to wrap that in a cache so we only build a `dict` every 5 minutes or so